### ISSUE TITLE
Add server->close() function

### ISSUE THREAD

**Issue by Kieobe** [Aircoookie/Espalexa#70]
Please add a function:
```
  void close()
  {
	server->close();
  }
```

Sometimes need to close the connection to use other libraries, but there is no such function. Thanks!

**Comment by Aircoookie** [comment]
Hi!
What library requires closing connections to work? We would also need a function to re-open the connection (basically restart the entire server).

Because Esplaexa is meant to be continuously reachable, this doesn't make too much sense. Can you tell me the interfering library? Maybe we can find a better solution :)

**Comment by Kieobe** [comment]
Library [https://github.com/tzapu/WiFiManager](WiFiManager) does not work (wifiManager.startConfigPortal()) until alexa connection is closed.

**Comment by anrada73** [comment]
it would be nice if you do it

**Comment by Aircoookie** [comment]
This works well for initial setup (open config AP if WiFi connection not configured or failed):

```cpp
void setup()
{
  WiFiManager wifiManager;
  wifiManager.autoConnect("Espalexa");

  //all the other Espalexa initialization
}
```

Does this satisfy your use case? If not, I will add the `close()` function if you need it.
The purpose of `startConfigPortal()` is to be able to reconfigure WiFi if already connected.
Espalexa is a "setup-and-forget" solution, so it would not make any sense at all to reconnect it to a different WiFi (while still in range of the previously configured WiFi)

**Comment by Kieobe** [comment]
Unfortunately, this option does not suit me, as I use the OnDemandConfigPortal.ino example, that is, launching the access point with the settings when the button is pressed. So I need a close() function. Since without closing the connection, the configuration page does not open. Thanks!

**Comment by SiddhantShah2001** [comment]
**Hey! @Aircoookie ** 
I'm having the same issue with my Project.
When i try to Set ESP into Config mode, it no Longer Responds.

This is my Code:
`#include <ESP8266WiFi.h>
#include <ESP8266WebServer.h>
#include <DNSServer.h>
#include <WiFiManager.h>
#include <EEPROM.h>

#include <Ticker.h>
Ticker ticker;

#define ESPALEXA_NO_SUBPAGE
#include <Espalexa.h>
Espalexa espalexa;

void Light(EspalexaDevice* dev);
EspalexaDevice* light;

const int on_off = 5; // D1
const int relay = 4;  // D2
const int red = 14;   // D5
const int green = 12; // D6

int on_off_state = LOW;
const int long_press = 7000;

int state;
unsigned int counter = 0;
uint8_t brightness;
int green_state = LOW;
const int state_address = 100;
unsigned int last_state;

const String ssid = WiFi.SSID().c_str();
const String pass = WiFi.psk().c_str();
int NCtag = LOW;
int Ctag = LOW;

void setup()
{
  EEPROM.begin(512);
  Serial.begin(74880);
  Serial.println("");
  Serial.println("Initialising!");

  last_state = EEPROM.read(state_address);
  state = last_state;

  if (state == HIGH)
  {
    brightness = 255;
  }
  else
  {
    brightness = 0;
  }

  Serial.print("State: ");
  Serial.println(state);

  pinMode(on_off, INPUT_PULLUP);
  pinMode(relay, OUTPUT);
  pinMode(red, OUTPUT);
  pinMode(green, OUTPUT);
  digitalWrite(relay,!state);

  WiFi.mode(WIFI_STA);
  if (ssid != NULL)
  {
    Serial.println("Credentials Found!");
    Serial.print("SSID: ");
    Serial.println(ssid);
    Serial.print("PASS: ");
    Serial.println(pass);

    WiFi.begin(ssid, pass);
  }
  else
  {
    Serial.println("Credentials not Found, Long Press Mode Button to Setup WiFi.");
  }

    light = new EspalexaDevice("LIGHT ND", Light, EspalexaDeviceType::onoff);
    espalexa.addDevice(light);
    light->setValue(brightness); 
    espalexa.begin();
}

void loop()
{
  if (WiFi.status() == WL_CONNECTED)
  {
    espalexa.loop(); //delay(1);
    ticker.detach();
    green_state = LOW;
    digitalWrite(green, HIGH);
  }
  if (WiFi.status() == WL_CONNECTED && Ctag == LOW)
  {
    Serial.println("Connected!");
    Serial.print("Connected to: ");
    Serial.print(WiFi.SSID().c_str());
    Serial.print(" Using Password: ");
    Serial.println(WiFi.psk().c_str());
    Serial.print("IP Address: ");
    Serial.println(WiFi.localIP());  
    Ctag = HIGH;
    NCtag = LOW;
  }
  else if (WiFi.status() != WL_CONNECTED && NCtag == LOW)
  {
    Serial.println("Not Connected");
    ticker.attach(0.2, nc_blink); 
    NCtag = HIGH;
    Ctag = LOW;
  }

  int read_on_off = digitalRead(on_off); 
  last_state = EEPROM.read(state_address);

  if (read_on_off == LOW)
  {
    counter++;
    on_off_state = HIGH;

    if (counter >= long_press)
    {
      Serial.println("Entering WiFi Setup Mode:");
      //counter = 0;
      long_pressed();
    }
    delay(1);
  }
  else if (read_on_off == HIGH && on_off_state == HIGH)
  {
    if (counter < long_press)
    {
      state = !state;
      EEPROM.put(state_address, state);
      EEPROM.commit();
      if (state == HIGH)
      {
        Serial.println("STATE: ON");
      }
      else
      {
        Serial.println("STATE: OFF");
      }
    }
    counter = 0;
    on_off_state = LOW;
    delay(250);
  }
  
  if (state == HIGH)
  {
     light->setValue(255);
     digitalWrite(red, HIGH);
     digitalWrite(relay, LOW);
  }
  else if(state == LOW)
  {
    light->setValue(0);
    digitalWrite(red, LOW);
    digitalWrite(relay, HIGH);
  }
}

void Light(EspalexaDevice* d)
{
  if (d == nullptr) return;

  if (d->getValue() > 0)
  {
    state = HIGH;
    EEPROM.write(state_address, state);
    EEPROM.commit();
    Serial.println("STATE: ON");
  }
  else
  {
    state = LOW;
    EEPROM.write(state_address, state);
    EEPROM.commit();
    Serial.println("STATE: OFF");
  }
}

void nc_blink()
{
  green_state = !green_state;
  digitalWrite(green, green_state);
}

void long_pressed()
{
  digitalWrite(red, LOW);
  digitalWrite(relay, HIGH);
  ticker.attach(0.1, nc_blink);
  WiFiManager wifiManager;
  wifiManager.setConfigPortalTimeout(120);
  wifiManager.startConfigPortal("NDLights", "smartmodules");
  ESP.reset();
}
`

Hope to see a Positive Reply!

**Comment by SiddhantShah2001** [comment]
When i add server->close() functon, Esp Crashes.

```
void long_pressed()
{
  ESP8266WebServer* server;
  server->close();
                           // It Crashes Here
  digitalWrite(red, LOW);
  digitalWrite(relay, HIGH);
  ticker.attach(0.1, nc_blink);
  WiFiManager wifiManager;
  wifiManager.setConfigPortalTimeout(120);
  wifiManager.startConfigPortal("NDLights", "smartmodules");
  ESP.reset();
}
```

Serial Monitor Output:

```
Exception (28):
epc1=0x40206199 epc2=0x00000000 epc3=0x00000000 excvaddr=0x00000014 depc=0x00000000

>>>stack>>>

ctx: cont
sp: 3fff01c0 end: 3fff0490 offset: 01a0
3fff0360:  ffffffff 3fffc6fc 00000000 3ffeee20  
3fff0370:  3ffeee48 3fffdad0 3ffeefd0 00000030  
3fff0380:  00000fe8 000001fd 000001fd 00000001  
3fff0390:  00000000 3ffeee68 3ffeecb8 40201d3e  
3fff03a0:  00000000 00000000 00000001 40100190  
3fff03b0:  3ffeed15 00000000 3fff114c 4020da98  
3fff03c0:  3ffe8d6a 00000000 4d000a0d 4020c3bd  
3fff03d0:  3ffe8d6a 00000000 6f43003a 4020c3bd  
3fff03e0:  3ffe8dad 00000000 3fff10d4 00000001  
3fff03f0:  40209738 3ffeeea0 3ffe8d68 40209744  
3fff0400:  40209738 3ffeeea0 3ffe8d68 40209a71  
3fff0410:  00000000 00000019 3ffeeea0 40209b68  
3fff0420:  00000000 3ffeee68 3ffeeea0 40209bdc  
3fff0430:  3fffdad0 3ffeee68 3ffeecb8 40204522  
3fff0440:  00000000 4e01a8c0 80000000 40209bdc  
3fff0450:  00000000 00000000 00000001 40100190  
3fff0460:  3fffdad0 00000000 3ffeef90 3ffeefd0  
3fff0470:  3fffdad0 00000000 3ffeef90 4020b41c  
3fff0480:  feefeffe feefeffe 3ffe84f8 40100cc1  
<<<stack<<<
```

**Comment by SiddhantShah2001** [comment]
**Okay, I got it to Work!**

### PROCEDURE:
**STEP 1:**
**Go to:** My PC > Documents > Arduino > libraries > Espalexa > src < and Open up all three Files in the Folder named as:
**"Espalexa.h"
"EspalexaDevice.h"
"EspalexaDevice.cpp"**
in a Simple Text Editor or a Code Editor (I used **Sublime Text**).

**STEP 2:**
In File: **"Espalexa.h":**
Add Line: 
```
void close()
  {
    server->close();
  }
```
  before the line:

  `~Espalexa(){delete devices;}`

In File: **"EspalexaDevice.h":**
Add Line:
```
void EspalexaDevice::close()
{
}
```

at the End, after the function:
```
void EspalexaDevice::doCallback()
{
  if (_callback != nullptr) {_callback(_val); return;}
  if (_callbackDev != nullptr) {_callbackDev(this); return;}
  if (_callbackCol != nullptr) _callbackCol(_val, getRGB());
}
```

In File: **"EspalexaDevice.cpp":**
Add Line:
`void close();`

Just after the Line:
`uint8_t getLastValue();`

**STEP 3:** Last Step!
Add Line:
`espalexa.close();`

just before you call the Config Portal.
Also Remember to add a reset function at the end or atleast the begin function for Espalexa.
i.e.:

```
  espalexa.close();
  WiFiManager wifiManager;
  wifiManager.startConfigPortal();
  ESP.reset(); // I would recommend this one than the one below.

or

  espalexa.close();
  WiFiManager wifiManager;
  wifiManager.startConfigPortal();
  espalexa.begin();
```

**That's It!**
